x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/ToCharFormatter.java

```diff
@@ -0,0 +1,371 @@
+                    "%03d",
+                    (t.get(WeekFields.ISO.weekOfWeekBasedYear()) - 1) * 7 + t.get(ChronoField.DAY_OF_WEEK))),
+            ToCharFormatter.of("DD").formatFn("d", x -> String.format(Locale.ROOT, "%02d", Integer.parseInt(x))),
+            ToCharFormatter.of("ID").formatFn("e"),
+            ToCharFormatter.of("D").formatFn(t -> String.valueOf(t.get(WeekFields.SUNDAY_START.dayOfWeek()))),
+            ToCharFormatter.of("W").formatFn(t -> String.valueOf(t.get(ChronoField.ALIGNED_WEEK_OF_MONTH))),
+            ToCharFormatter.of("WW").formatFn(t -> String.format(Locale.ROOT, "%02d", t.get(ChronoField.ALIGNED_WEEK_OF_YEAR))),
+            ToCharFormatter.of("IW").formatFn(t -> String.format(Locale.ROOT, "%02d", t.get(WeekFields.ISO.weekOfWeekBasedYear()))),
+            ToCharFormatter.of("CC").formatFn(t -> {
+                int century = century(t.get(ChronoField.YEAR));
+                return String.format(Locale.ROOT, century < 0 ? "%03d" : "%02d", century);
+            }),
+            ToCharFormatter.of("J").formatFn("g", modifiedJulianDay -> String.valueOf(Integer.parseInt(modifiedJulianDay) + 2400001)),
+            ToCharFormatter.of("Q").formatFn("Q"),
+            ToCharFormatter.of("RM").formatFn("MM", month -> String.format(Locale.ROOT, "%-4s", toRome(Integer.parseInt(month)))).text(),
+            ToCharFormatter.of("rm")
+                .formatFn("MM", month -> String.format(Locale.ROOT, "%-4s", toRome(Integer.parseInt(month)).toLowerCase(Locale.ROOT)))
+                .text(),
+            ToCharFormatter.of("TZ").formatFn(ToCharFormatter::zoneAbbreviationOf).text(),
+            ToCharFormatter.of("tz").formatFn(t -> zoneAbbreviationOf(t).toLowerCase(Locale.ROOT)).text(),
+            ToCharFormatter.of("TZH").lowercaseAccepted(false).formatFn("ZZ", s -> s.substring(0, 3)).text(),
+            ToCharFormatter.of("TZM").lowercaseAccepted(false).formatFn("ZZ", s -> lastNCharacter(s, 2)).text(),
+            ToCharFormatter.of("OF")
+                .lowercaseAccepted(false)
+                .formatFn("ZZZZZ", ToCharFormatter::formatOffset)
+                .text(ToCharFormatter::fillOffset)).stream().map(Builder::build).collect(Collectors.toList());
+
+        Map<String, ToCharFormatter> formatterMap = new LinkedHashMap<>();
+        for (ToCharFormatter formatter : formatters) {
+            formatterMap.put(formatter.pattern, formatter);
+        }
+        // also index the lower case version of the patterns if accepted
+        for (ToCharFormatter formatter : formatters) {
+            if (formatter.lowercaseAccepted) {
+                formatterMap.putIfAbsent(formatter.pattern.toLowerCase(Locale.ROOT), formatter);
+            }
+        }
+        FORMATTER_MAP = formatterMap;
+    }
+
+    @SuppressWarnings("OptionalGetWithoutIsPresent") private static final int
+        MAX_TO_CHAR_FORMAT_STRING_LENGTH =
+        FORMATTER_MAP.keySet().stream().mapToInt(String::length).max().getAsInt();
```

Comment:
True, but we can spare the cycles since this is a static number, can be computed once, the list of patterns does not change.

Note: will put the annotation into a new line.